File: src/main/java/com/dh/ondot/schedule/application/dto/HomeScheduleListItem.java
package com.dh.ondot.schedule.application.dto;

import com.dh.ondot.core.util.TimeUtils;
import com.dh.ondot.schedule.api.response.AlarmDto;
import com.dh.ondot.schedule.domain.Schedule;

import java.time.*;
import java.util.List;

public record HomeScheduleListItem(
        Long scheduleId,
        Double startLongitude,
        Double startLatitude,
        Double endLongitude,
        Double endLatitude,
        String scheduleTitle,
        boolean isRepeat,
        List<Integer> repeatDays,
        LocalDateTime appointmentAt,
        AlarmDto preparationAlarm,
        AlarmDto departureAlarm,
        boolean hasActiveAlarm
) {
    public static HomeScheduleListItem from(Schedule schedule) {
        return new HomeScheduleListItem(
                schedule.getId(),
                schedule.getDeparturePlace().getLongitude(),
                schedule.getDeparturePlace().getLatitude(),
                schedule.getArrivalPlace().getLongitude(),
                schedule.getArrivalPlace().getLatitude(),
                schedule.getTitle(),
                schedule.getIsRepeat(),
                schedule.getRepeatDays() == null ? List.of() : List.copyOf(schedule.getRepeatDays()),
                TimeUtils.toSeoulDateTime(schedule.getAppointmentAt()),
                AlarmDto.of(schedule.getPreparationAlarm(), schedule),
                AlarmDto.of(schedule.getDepartureAlarm(), schedule),
                schedule.hasAnyActiveAlarm()
        );
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/application/dto/HomeScheduleListItem.kt
package com.dh.ondot.schedule.application.dto

import com.dh.ondot.core.util.TimeUtils
import com.dh.ondot.schedule.presentation.response.AlarmDto
import com.dh.ondot.schedule.domain.Schedule
import java.time.Instant
import java.time.LocalDateTime

data class HomeScheduleListItem(
    val scheduleId: Long?,
    val startLongitude: Double?,
    val startLatitude: Double?,
    val endLongitude: Double?,
    val endLatitude: Double?,
    val scheduleTitle: String?,
    val isRepeat: Boolean,
    val repeatDays: List<Int>,
    val appointmentAt: LocalDateTime?,
    val preparationAlarm: AlarmDto?,
    val departureAlarm: AlarmDto?,
    val hasActiveAlarm: Boolean,
    val preparationNote: String?,
    val nextAlarmAt: Instant?,
) {
    companion object {
        fun from(schedule: Schedule): HomeScheduleListItem {
            return HomeScheduleListItem(
                scheduleId = schedule.id,
                startLongitude = schedule.departurePlace!!.longitude,
                startLatitude = schedule.departurePlace!!.latitude,
                endLongitude = schedule.arrivalPlace!!.longitude,
                endLatitude = schedule.arrivalPlace!!.latitude,
                scheduleTitle = schedule.title,
                isRepeat = schedule.isRepeat,
                repeatDays = if (schedule.repeatDays == null) listOf() else schedule.repeatDays!!.toList(),
                appointmentAt = TimeUtils.toSeoulDateTime(schedule.appointmentAt),
                preparationAlarm = AlarmDto.of(schedule.preparationAlarm!!, schedule),
                departureAlarm = AlarmDto.of(schedule.departureAlarm!!, schedule),
                hasActiveAlarm = schedule.hasAnyActiveAlarm(),
                preparationNote = schedule.preparationNote,
                nextAlarmAt = schedule.calculateNextAlarmAt(),
            )
        }
    }
}
